File: Project Οντοκεντρικός Προγραμματισμός/Java Project/Hotel Project, Swing/Hotel.java
import java.util.*;
public class Hotel
{

    private String hotelName;
    protected ArrayList<Room> hotelRoom = new ArrayList<Room>(1); //Λίστα με τα δωμάτια.
    protected ArrayList<Booking> hotelBooking = new ArrayList<Booking>(1); //Λίστα με τις κρατήσεις.
    //Getter
    public String getHotelName()
    {
        return hotelName;
    }
   
    public Hotel(String name)//Constructor για το αντικείμενο Hotel.
    {
        hotelName=name;
    }
    
    public void addHotelRoom(Room room)//Μέθοδος προσθήκης δωματίου στη λίστα δωματίων.
    {
            hotelRoom.add(room);
    }
    public Room recoverRoomOutOfID(int ID)//Μέθοδος η οποία βρίσκει το δωμάτιο με δεδομένο ID. 
    {
        int i;
        for(i=0; i<hotelRoom.size(); i++)
        {
            if(hotelRoom.get(i).getRoomID()==ID)
            {   
                return hotelRoom.get(i);
            }
        }
        return null;  
    }
    public Booking recoverBookingOutOfID(int ID)//Μέθοδος η οποία βρίσκει τη κράτηση με δεδομένο ID.
    {
        int i;
        for(i=0; i<hotelBooking.size(); i++)
        {
            if(hotelBooking.get(i).getBookingID()==ID)
            {   
                return hotelBooking.get(i);
            }
        }
        return null;      
    }
    public boolean addBookingToRoom(Booking aa,int ID)//Μέθοδος προσθήκης κράτησης σε ένα δωμάτιο.
    {
       Room roomaki = new Room();
       int i;
       if(recoverRoomOutOfID(ID)==null || roomaki.addBooking(aa)==false) //Αν δεν υπάρχει δωμάτιο ή το δωμάτιο έχει ήδη κρατηθεί.
       {
           System.out.println("Booking not successful.");
           return false;
       }
       if(recoverRoomOutOfID(ID) instanceof RoomTypeC && (aa.getPeopleNum()<RoomTypeC.getMinPeople() || aa.getAccomodationDays()<RoomTypeC.getMinDays()))//Αν δεν ταιριάζουν τα ελάχιστα άτομα ή οι ελάχιστες μέρες για κράτηση σε δωμάτιο τύπου C ή E.
       {
           System.out.println("Booking not successful, cannot Book in Room Type C not enough days or not enough people.");
           return false;           
       }
       if(recoverRoomOutOfID(ID) instanceof RoomTypeD && (aa.getAccomodationDays()>1))//Αν ο χρήστης δηλώσει παραπάνω μέρες από μία για να γίνει κράτηση σε δωμάτιο τύπου D.
       {
           System.out.println("Booking not successful, cannot Book in Room Type D single day room.");
           return false; 
       }
       recoverRoomOutOfID(ID).addBooking(aa); 
       hotelBooking.add(aa);
       System.out.println("Booking succcessful with ID: " + aa.getBookingID());
       return true;
    }
    public int addBookingNoID(Booking aaa)//Μέθοδος προσθήκης κράτησης σε δωμάτιο με βάση το ID.
    {
        int i;
        boolean found=false;
        for(i=0; i<hotelRoom.size(); i++)
        {
            Room roomaki = new Room(); //Το αντικείμενο Roomaki πρέπει να αρχικοποιείται σε κάθε επανάληψη.
            if( roomaki.addBooking(aaa)==false) 
            {
                System.out.println("Booking not Succesful.");
                return 0;
            }
            else if(aaa.getAccomodationDays()==1)
            {
                if(hotelRoom.get(i) instanceof RoomTypeD)
                {
                    if(hotelRoom.get(i).addBooking(aaa)==true)
                       System.out.println("Booking with ID: " + aaa.getBookingID() + " Successfully booked in " + aaa.getRoom().getClass());
                    else
                        continue;
                    hotelBooking.add(aaa);
                    found=true;
                    break;
                }
                else    
                    continue;
            }
            else if(aaa.getPeopleNum() >= RoomTypeC.getMinPeople() && aaa.getAccomodationDays() >= RoomTypeC.getMinDays())
            {
                if(hotelRoom.get(i) instanceof RoomTypeC  || hotelRoom.get(i) instanceof RoomTypeE)
                {
                    if(hotelRoom.get(i).addBooking(aaa)==true) 
                        System.out.println("Booking with ID: " + aaa.getBookingID() + " Successfully booked in " + aaa.getRoom().getClass());
                    else 
                        continue;
                    hotelBooking.add(aaa);
                    found=true;
                    break;
                }
                else
                    continue;
            }
            else
            {
                if(hotelRoom.get(i).addBooking(aaa)==true)
                    System.out.println("Booking with ID: " + aaa.getBookingID() + " Successfully booked in " + aaa.getRoom().getClass());
                else
                    continue;
                hotelBooking.add(aaa);
                found=true;
                break;
            }
          
        }
        if(found==true)
        {
            if(i == hotelRoom.size())
                i = (i - 1 );
            System.out.println("Room ID: " + hotelRoom.get(i).getRoomID());
            return hotelRoom.get(i).getRoomID();
        }
        else
        {   
            System.out.println("Not available Booking");
            return 0;
        }
    }
    public void deleteBooking(int ID) //Μέθοδος διαγραφής κράτησης σε ένα δωμάτιο με βάση το ID.
    {
        if(recoverBookingOutOfID(ID)!=null)        
        {
          if(recoverBookingOutOfID(ID).getRoom().cancelBooking(ID)==true)
           {
               recoverBookingOutOfID(ID).getRoom().cancelBooking(ID);
               hotelBooking.remove(recoverBookingOutOfID(ID));
               System.out.println("Cancelation Successful!");
           }
        }
        else
        {
            System.out.println("There is not a booking with this ID, check bookings and try again.");
        }
    }
    public double income(int ID)//Μέθοδος υπολογισμού των εσόδων ενός δωματίου με βάση το ID.
    {
        if(recoverRoomOutOfID(ID)==null)
        {
            System.out.println("No room with this ID");    
            return 0;
        }
        System.out.println("Money earned from this room: " + recoverRoomOutOfID(ID).getPriceOfRoom());
        return recoverRoomOutOfID(ID).getPriceOfRoom();
    }
    public double income()//Μέθοδος υπολογισμού των εσόδων ενός δωματίου.
    {
        double totalEarnings=0;
        int i;
        for(i=0; i < hotelRoom.size(); i++)
        {
            totalEarnings += hotelRoom.get(i).getPriceOfRoom(); 
        }
        return totalEarnings;
    }
    public void bookingPlan()//Μεθοδος εκτύπωσης του πλάνου κρατήσεων του ξενοδοχείου.
    {
        int i,k;
        System.out.print("Room\t\t");
        for(i=1; i<31; i++)
        {
            System.out.print(i +"\t");
        }
        System.out.println("\n");
        for(i=0; i < hotelRoom.size(); i++)
        {
            System.out.print(hotelRoom.get(i).getRoomID() + "\t\t");
            for(k=1; k<31; k++)
            {
                if(hotelRoom.get(i).roomAvailability[k]==null)
                {
                    System.out.print("_" + "\t");
                }
                else
                {
                    System.out.print("*" + "\t");
                }
            }
            System.out.println();
        }
    }
}

File: Project Οντοκεντρικός Προγραμματισμός/C++ Project/Visual Studio Hotel Project/ConsoleApplication1/Hotel.cpp
#include "Hotel.h"
#include <vector>
#include <algorithm>

string hotelName;//Variable used for the name of a Hotel type object.
vector<Room> hotelRoom;//Vector used for the rooms of a Hotel type object.
vector<Booking> hotelBooking;//Vector used for the bookings of a Hotel type object.


Hotel::Hotel(string name)//Constructor for a Hotel type object.
{
	hotelName = name;
	vector<Room*> hotelRoom;//Creation of vector used for Room type pointers.
	vector<Booking*> hotelBooking;//Creation of vector used for Booking type pointers.
}

Hotel::~Hotel()//Default destructor for a Hotel type object.
{
}


bool operator>(Booking& bb, RoomTypeC& cc)//Implementation for the overloading of the operator >, used by a Booking and RoomTypeC type objects.
{
	return (bb.getPeopleNum() < cc.getMinPeople() || bb.getAccomodationDays() < cc.getMinDays());//Checking the minDays and the minPeople variables of a Room *bb and a RoomTypeC *cc.
}

bool operator<(int x, Booking& booking)//Implementation for the overloading of the operator <, used by a Booking type object and an integer.
{
	return x < booking.getAccomodationDays();//Checking if the accomodationDays variable, of Booking *bb, is greater than x.
}

void Hotel::addHotelRoom(Room* room)//Function that puts a Room type object in the hotelRoom vector.
{
	hotelRoom.push_back(room);
}

Room* Hotel::recoverRoomOutOfID(int ID)//Function that returns a Room type pointer if there is a room with the ID or 0(null pointer) if not.
{
	unsigned int i;
	for (i = 0; i<hotelRoom.size(); i++)
	{
		if (hotelRoom[i]->getRoomID() == ID)
		{
			return hotelRoom[i];
		}
	}
	return false;
}

Booking* Hotel::recoverBookingOutOfID(int ID)//Function that returns a Booking type pointer if there is a booking with the ID or 0(null pointer) if not.
{
	unsigned int i;
	for (i = 0; i < hotelBooking.size(); i++)
	{
		if (hotelBooking[i]->getBookingID() == ID)
			return hotelBooking[i];
	}
	return false;
}

bool Hotel::addBookingToRoom(Booking* aa, int ID)//Function that returns true if you can book in a room with the ID or false if not.
{
	Room *roomaki = new Room();
	if ((roomaki->addBooking(aa) == false) || (recoverRoomOutOfID(ID)==false))
	{
		cout << "Booking not successful" << endl;
		return false;
	}
	if (RoomTypeE* rte = dynamic_cast<RoomTypeE*>(recoverRoomOutOfID(ID)))//Checking if the room *aa is a RoomTypeC object.
	{
		if (*aa > *rte)
		{
			cout << "Booking not successful, cannot Book in Room Type E not enough days or not enough people." << endl;
			return false;
		}
	}
	if (RoomTypeC* rtc = dynamic_cast<RoomTypeC*>(recoverRoomOutOfID(ID)))//Checking if the room *aa is a RoomTypeC object.
	{
		if (*aa > *rtc)
		{
			cout << "Booking not successful, cannot Book in Room Type C not enough days or not enough people." << endl;
			return false;
		}
	}
	if (RoomTypeD* rtd = dynamic_cast<RoomTypeD*>(recoverRoomOutOfID(ID)))//Checking if the room *aa is a RoomTypeD object.
	{
		if (1 < *aa)
		{
			cout << "Booking not successful, cannot Book in Room Type D single day room." << endl;
			return false;
		}

	}
	recoverRoomOutOfID(ID)->addBooking(aa);
	hotelBooking.push_back(aa);
	cout << "Booking succcessful with ID: " << aa->getBookingID() << endl;
	return true;
}

int Hotel::addBookingNoID(Booking* aaa)//Function that returns true if you can book in a room or false if not.
{
	unsigned int i;
	for (i = 0; i < hotelRoom.size(); i++)
	{
		if (hotelRoom[i]->addBooking(aaa) == true)
		{
			hotelRoom[i]->addBooking(aaa);
			cout << "Booking successful in room with id: " << hotelRoom[i]->getRoomID() << endl;
			cout << "Booking ID: " << aaa->getBookingID() <<endl;
			hotelBooking.push_back(aaa);
			return hotelRoom[i]->getRoomID();
		}

	}
	cout << "NOT SUCCESSFUL BOOKING." << endl;
	return 0;
}

void Hotel::deleteBooking(int ID)//Function that returns a message that informs if you can delete a booking or if there isn't a booking with an ID.
{
	if (recoverBookingOutOfID(ID) != 0)
	{
		if (recoverBookingOutOfID(ID)->getRoom()->cancelBooking(ID) == true)
		{
			recoverBookingOutOfID(ID)->getRoom()->cancelBooking(ID);
			hotelBooking.erase(remove(hotelBooking.begin(), hotelBooking.end(), recoverBookingOutOfID(ID)), hotelBooking.end());//Searches the whole vector and if it finds the object we want, it will get deleted.
			cout << "Cancelation Successful!" << endl;
		}
	}
	else
		cout << "There is not a booking with this ID, check bookings and try again." << endl;
}

double Hotel::income(int ID)//Function that returns the amount of income earned from a room.
{
	if (recoverRoomOutOfID(ID) == false)
	{
		cout << "No room with this ID" << endl;
		return 0;
	}
	cout << "Money earned from this room " << recoverRoomOutOfID(ID)->getPriceOfRoom() << "." << endl;
	return recoverRoomOutOfID(ID)->getPriceOfRoom();
}

double Hotel::income()//Function that returns the amount of income earned from the hotel.
{
	double totalEarnings = 0;
	unsigned int i;
	for (i = 0; i < hotelRoom.size(); i++)
	{
		totalEarnings += hotelRoom[i]->getPriceOfRoom();
	}
	return totalEarnings;
}

void Hotel::bookingPlan()//Function that prints the booking plan of all the rooms currently at the hotel.
{
	unsigned int i, k;
	cout << "Room\t\t";
	for (i = 1; i<31; i++)
	{
		cout << i << "\t";
	}
	cout << endl;
	for (i = 0; i < hotelRoom.size(); i++)
	{
		cout << hotelRoom[i]->getRoomID() << "\t\t";
		for (k = 1; k<31; k++)
		{
			if (hotelRoom[i]->getRoomAvailability(k) == 0)
			{
				cout << "_\t";
			}
			else
			{
				cout << "*\t";
			}
		}
		cout << endl;
	}
}

string getHotelName()//Getter for the variable hotelName.
{ 
	return hotelName; 
}

void setHotelName(string name)//Setter for the variable hotelName.
{
	hotelName = name;
}